Dao wallet

matt-o-how opened the pull request:
### Purpose:
Merge DAO Wallet into main

### Testing Notes:
tests/wallet/dao_wallet

aqk committed:
DAO Wallet changes

aqk committed:
Newly created DAO Files

matt-o-how committed:
add assertion to prevent voting with 0 value

aqk committed:
Take changes from dao_unlock_cat_fix

aqk committed:
Bring in chia/wallet/wallet_state_manager.py changes from dao_unlock_cat_fix

aqk committed:
Bring chia/wallet/wallet_state_manager.py up to date with main

aqk committed:
Merge branch 'main' into dao-wallet

aqk committed:
Typechecking and function input validation

aqk committed:
linting

geoffwalmsley committed:
add singleton aggregator back into branch

matt-o-how committed:
add singleton aggregator back into branch (#15662)

matt-o-how committed:
fix dao_wallet bug and commit hex files

matt-o-how committed:
re-add amount check assertion to proposal voting

matt-o-how committed:
delete unused tests

geoffwalmsley committed:
remove unused dao_util funcs

geoffwalmsley committed:
move singleton funcs from dao_utils to singleton.py

matt-o-how committed:
add concurrency test to check wallet isn't locking up coins when spend fails

matt-o-how committed:
fix API bug and remove unused clsphex

geoffwalmsley committed:
spend timer coin during proposal self destruct

geoffwalmsley committed:
add proposal_minimum_amount to CLI

geoffwalmsley committed:
remove debug statements from dao_lockup

geoffwalmsley committed:
add hints for p2_singletons and brute-force the treasury balance

geoffwalmsley committed:
return proposal id when creating new proposals

geoffwalmsley committed:
improve consistency of xch / mojo units in CLI

geoffwalmsley committed:
Revert "spend timer coin during proposal self destruct"

This reverts commit e5ec333d6266eefb453adf246e992543fd91f5a7.

matt-o-how committed:
DAO wallet todos (#15682)

- cleanup dao utils and move get_singleton_struct to singleton.py
- spend the timer coin during proposal self destruct

matt-o-how committed:
change lockup to use self-hash and add in a test for coin removal

matt-o-how committed:
prep for moving proposal to 2 stage curry

geoffwalmsley committed:
add self destruct option to CLI and RPC

geoffwalmsley committed:
remove a breakpoint

geoffwalmsley committed:
fix partial vote on proposal creation

geoffwalmsley committed:
add a function to get the treasury id

geoffwalmsley committed:
clarify the blocks until closable numbers

matt-o-how committed:
begin move to add proposal self hash
down to only one mismatched assert

geoffwalmsley committed:
remove redundant else

matt-o-how committed:
CLI fixes for DAO wallets (#15692)

geoffwalmsley committed:
clean up dao creation funcs

matt-o-how committed:
fix puzzle announcement mismatch and further optimise dao_lockup.clsp

matt-o-how committed:
wallet fixes

geoffwalmsley committed:
reorder standard wallet funcs and remove is_spend_retrievable

matt-o-how committed:
fix proposal timer coin creation

matt-o-how committed:
another fix

geoffwalmsley committed:
refactoring

geoffwalmsley committed:
remove old comments and breakpoints

geoffwalmsley committed:
Merge branch 'dao-wallet' into gw.dao_wallet_cleanup

matt-o-how committed:
split out singleton struct across the two currys

matt-o-how committed:
fix sync error

matt-o-how committed:
fix bad spendbundle reveal

matt-o-how committed:
fix remaining bugs

geoffwalmsley committed:
revert get_balance to using the node to find coins

geoffwalmsley committed:
fix missing self_destruct in rpc and test

matt-o-how committed:
dao wallet cleanup (#15753)

tidy up and some refactoring of dao_wallet.py

matt-o-how committed:
Merge branch 'dao-wallet' into dao_curry_twice

matt-o-how committed:
remove debug (x) from proposal_validator

matt-o-how committed:
move comment for clarification

matt-o-how committed:
move to await wallet sync instead of sleep in loop

matt-o-how committed:
remove long block farming

matt-o-how committed:
Dao curry twice (#15773)

### Purpose:
Optimise the lockup and proposal chialisp files to use the curry twice
technique, making the cost of calculating their puzzle hashes
dramatically lower elsewhere in code.

Also fixed a couple of other minor bugs while I was in there.

matt-o-how committed:
remove unused clvm

trepca commented on chia/cmds/dao_funcs.py:
```diff
@@ -0,0 +1,456 @@
+        votes_against = prop["amount_voted"] - prop["yes_votes"]
+        print(f"Votes against: {votes_against}")
+        print("------------------------")
+    print(f"Proposals have {soft_close_length} blocks of soft close time.")
```
what would `soft_close_length` be if proposal is already closed?

trepca commented on chia/cmds/dao_funcs.py:
```diff
@@ -0,0 +1,456 @@
+        name = res["tx_id"]
+        print(f"Submitted proposal close transaction with name: {name}")
+    else:
+        print("Unable to generate close transaction.")
```
could we provide more info here on why it's not able to close?

matt-o-how committed:
remove unused files and partial fix clvm tests

geoffwalmsley committed:
fix clvm tests

matt-o-how committed:
update dao files listed in deployed_puzzle_hashes.json

trepca commented on chia/wallet/cat_wallet/cat_wallet.py:
```diff
@@ -176,6 +176,7 @@ async def create_new_cat_wallet(
         chia_tx = dataclasses.replace(chia_tx, spend_bundle=spend_bundle)
         await self.standard_wallet.push_transaction(chia_tx)
         await self.standard_wallet.push_transaction(cat_record)
+        # breakpoint()
```
let remove this

trepca commented on chia/wallet/cat_wallet/cat_wallet.py:
```diff
@@ -806,7 +807,7 @@ async def generate_unsigned_spendbundle(
         )
 
-    async def generate_signed_transaction(
+    async def generate_signed_transactions(
```
@Quexington is this ok? to rename this

trepca commented on chia/wallet/cat_wallet/cat_wallet.py:
```diff
@@ -820,6 +821,7 @@ async def generate_signed_transaction(
         max_coin_amount: Optional[uint64] = None,
         excluded_coin_amounts: Optional[List[uint64]] = None,
         reuse_puzhash: Optional[bool] = None,
+        override_memos: Optional[bool] = None,
```
why do we need this? this functions already accepts memos?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    treasury_id: bytes32, xch_conditions: Optional[List], asset_conditions: Optional[List[Tuple]]  # type: ignore
+) -> Program:
+    # TODO: typecheck get_spend_p2_singleton_puzzle arguments
+    # TODO: add tests for get_spend_p2_singleton_puzzle: pass xch_conditions as Puzzle, List and ConditionWithArgs
```
has this been done? (the todos)

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    return puzzle
+
+
+def add_proposal_to_active_list(
```
what's an active list? can we document this function?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    proposed_puzzle_hash: bytes32,
+) -> Program:
+    """
+    spend_or_update_flag can take on the following values, ranked from safest to most dangerous:
```
spend_or_update_flag? where is that?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+        PAYOUT_PUZHASH,
+    ) = curried_args.as_iter()
+    return DAORules(
+        uint64(int_from_bytes(proposal_timelock.as_atom())),
```
use `as_int()` instead of `as_atom` here

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    )
+
+
+# This takes the treasury puzzle and treasury solution, not the full puzzle and full solution
```
lets make this as docstring so can be discoverable

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    #     proposal_self_destruct_length,
+    #     oracle_spend_delay,
+    # ) = curried_args
+    if solution.rest().rest().first() != Program.to(0):
```
`solution.at("rrf")`

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+# This also returns the proposal puzzle and not the full puzzle
+def get_new_puzzle_from_proposal_solution(puzzle_reveal: Program, solution: Program) -> Optional[Program]:
+    # Check if soft_close_length is in solution. If not, then add votes, otherwise close proposal
+    if len(solution.as_python()) == 1:
```
`.as_python` will be deprecated as can cause DOS attack, can we use something else here?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    return puzzle
+
+
+# TODO: move curry_singleton to chia.wallet.singleton
```
have these TODOs been done?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    except ValueError:
+        import traceback
+
+        print(f"exception: {traceback.format_exc()}")
```
why do we print here?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    return None
+
+
+# This is for use in the WalletStateManager to determine the type of coin received
```
docstring

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+    return None
+
+
+# This is used in WSM to determine whether we have a dao funding spend
```
docstring

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+
+# This is used in WSM to determine whether we have a dao funding spend
+def match_funding_puzzle(uncurried: UncurriedPuzzle, solution: Program) -> Optional[bool]:
+    # TODO: handle case where solution is for existing p2_singleton
```
completed?

trepca commented on chia/wallet/dao_wallet/dao_utils.py:
```diff
@@ -0,0 +1,763 @@
+def match_funding_puzzle(uncurried: UncurriedPuzzle, solution: Program) -> Optional[bool]:
+    # TODO: handle case where solution is for existing p2_singleton
+    try:
+        if match_cat_puzzle(uncurried):
```
can we comment what's happening in this if ?

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+        )
+        self.wallet_id = self.wallet_info.id
+        std_wallet_id = self.standard_wallet.wallet_id
+        bal = await wallet_state_manager.get_confirmed_balance_for_wallet(std_wallet_id)
```
why do we fetch balance again here?

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+        self.log = logging.getLogger(name if name else __name__)
+        std_wallet_id = self.standard_wallet.wallet_id
+        bal = await wallet_state_manager.get_confirmed_balance_for_wallet(std_wallet_id)
+        if amount_of_cats > bal:
```
1 cat = 1 mojo in this case?

geoffwalmsley replied:
yes, amount_of_cats is in mojos.

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+        await self.wallet_state_manager.add_new_wallet(self)
+        await self.save_info(self.dao_info)
+        if self.wallet_info is None:
+            raise ValueError("Internal Error")
```
why would this happen?

geoffwalmsley replied:
good point. The create_wallet method will raise a ValueError if wallet_info is none anyway, so this check is redundant.

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+    def id(self) -> uint32:
+        return self.wallet_info.id
+
+    def puzzle_for_pk(self, pubkey: G1Element) -> Program:
```
why do we need this method since it's a constant?

geoffwalmsley replied:
I thought these were necessary for wallet integration, but seems like they aren't so have removed `puzzle_for_pk`, `puzzle_hash_for_pk`, and `get_new_puzzle`

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+    async def get_frozen_amount(self) -> uint64:
+        return uint64(0)
+
+    async def get_max_send_amount(self, records: Optional[Set[WalletCoinRecord]] = None) -> uint128:
```
so DAO can't be transferred to a different wallet?

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+        This can be either a treasury coin update or funds added to the treasury
+        """
+        self.log.info(f"DAOWallet.coin_added() called with the coin: {coin.name()}:{coin}.")
+        wallet_node: Any = self.wallet_state_manager.wallet_node
```
why any and not WalletNode?

geoffwalmsley replied:
Importing WalletNode in wallets creates a circular dependency via WSM. So has to be either untyped or Any

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+                    dao_info = dataclasses.replace(self.dao_info, assets=new_asset_list)
+                    await self.save_info(dao_info)
+        except Exception as e:
+            self.log.error(f"Error occurred during dao wallet coin addition: {e}")
```
log.exception

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+        if peer is None:
+            raise ValueError("Could not find any peers to request puzzle and solution from")
+
+        children = await wallet_node.fetch_children(parent_coin_id, peer)
```
why do you fetch children here?
you could just have `while True` loop and break as you are breaking now in loop below?

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+        await self.save_info(dao_info)
+        return
+
+    async def resync_treasury_state(self) -> None:
```
not sure what this does, please add a docstring

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+                return get_curry_vals_from_proposal_puzzle(prop.inner_puzzle)
+        raise ValueError("proposal not found")
+
+    # This has to be in the wallet because we are taking an ID and then searching our stored proposals for that ID
```
duplicate of the above, not sure why it's needed

trepca commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+            ],  # create cat_launcher coin
+            [
+                60,
+                Program.to(["m", full_puz.get_tree_hash()]).get_tree_hash(),
```
I assume `m` is mint? can we make it into an enum so it's clear?

trepca commented on tests/wallet/dao_wallet/test_dao_clvm.py:
```diff
@@ -0,0 +1,973 @@
+
+
+def test_validator() -> None:
+    # This test covers proposal_validator and spend_p2_singleton
```
not sure what this means, what test cases and scenarios does this cover?

trepca commented on tests/wallet/dao_wallet/test_dao_clvm.py:
```diff
@@ -0,0 +1,973 @@
+    return
+
+
+def test_treasury() -> None:
```
hard to see which scenarios we're testing here, can you describe here please or create separate tests for each

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    for i in range(1, 11):
+        await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(puzzle_hash_0))
+    if not trusted:
+        await asyncio.sleep(1)
```
?

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+
+    # Give the full node a moment to catch up if there are no trusted peers
+    if not trusted:
+        await asyncio.sleep(1)
```
this will make this test flaky, wait for something instead of doing sleep

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    new_inner_puzhash = await dao_wallet_0.get_new_p2_inner_hash()
+    assert isinstance(new_inner_puzhash, bytes32)
+
+    # # generate new dao spends with bad cat details
```
why is this commented out?

geoffwalmsley replied:
We removed the functions for creating DAO's for existing CATs and creating DAOs 'offline' via spendbundle, since they were out of scope for the beta. These tests were relating to those functions, so I've just removed it for now. We've stored the unused code in a WIP branch so it can be picked up again later.

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    funding_tx = await dao_wallet_0.create_add_money_to_treasury_spend(xch_funds)
+    funding_sb = funding_tx.spend_bundle
+    await time_out_assert_not_none(5, full_node_api.full_node.mempool_manager.get_spendbundle, funding_sb.name())
+    await full_node_api.process_transaction_records(records=[funding_tx])
```
why not process cat and xch funding in one block here?

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    for i in range(1, 11):
+        await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(puzzle_hash_0))
+
+    # async with wallet_node_0.wallet_state_manager.lock:
```
can we remove this?

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    await full_node_api.process_spend_bundles(bundles=[proposal_sb])
+
+    # Give the wallet nodes a second
+    await asyncio.sleep(1)
```
why a second exactly? isn't there any other way to detecting that nodes have been updated?

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    for i in range(1, num_blocks):
+        await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(puzzle_hash_0))
+    # Wait a second for the dao cat wallet to update
+    await asyncio.sleep(2)
```
2 seconds here?

trepca commented on tests/wallet/dao_wallet/test_dao_wallets.py:
```diff
@@ -0,0 +1,2067 @@
+    for i in range(1, 11):
+        await full_node_api.farm_new_transaction_block(FarmNewBlockProtocol(puzzle_hash_0))
+
+    # async with wallet_node_0.wallet_state_manager.lock:
```
can we remove this commented block?

geoffwalmsley commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+
+        # Resync the wallet from when the treasury was created to get the existing funds
+        # TODO: Maybe split this out as an option for users since it may be slow?
+        if not wallet_node.is_trusted(peer):
```
Untrusted nodes won't automatically send us the history of all the treasury and proposal coins, so we have to request all the updates between the treasury being created and current peak.

geoffwalmsley committed:
Address sebs comments for dao_wallet

kimsk committed:
fix error in dao_get_treasury_balance RPC

geoffwalmsley committed:
add docstrings and comments to clvm tests

geoffwalmsley committed:
fix timings in dao_wallet_tests

trepca committed:
fix error in dao_get_treasury_balance RPC (#15818)

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
fixes for wsm, wallet, rpc after merging main

kimsk committed:
remove execute_with_wallet and use get_wallet_client

geoffwalmsley committed:
pre-commit fixes

geoffwalmsley committed:
bring trade manager up to date with main

geoffwalmsley committed:
Merge branch 'dao-wallet' into karlkim.dao-cmds

geoffwalmsley committed:
dao-wallet -- fixes dao cmds, remove execute_with_wallet and use get_wallet_client (#15837)

### Purpose:
After the commit
https://github.com/Chia-Network/chia-blockchain/commit/78e499107d79cbbe92c7c3dea3d946972a822ef4
(merging main), the dao cmds are not working anymore because
`execute_with_wallet` function has been removed from
[cmds_util.py](https://github.com/Chia-Network/chia-blockchain/commit/78e499107d79cbbe92c7c3dea3d946972a822ef4#diff-1e43439572b349526b913914815647ba42e0d57b86edb4f0e18e1ea6d812033c)
and `get_wallet_client` has been created.


### Current Behavior:
```
❯ chia dev sim --root-path $CHIA_SIM_ROOT start -wr
Traceback (most recent call last):
  File "/Users/karlkim/chia-blockchain/venv/bin/chia", line 5, in <module>
    from chia.cmds.chia import main
  File "/Users/karlkim/chia-network/chia-blockchain/chia/cmds/chia.py", line 12, in <module>
    from chia.cmds.dao import dao_cmd
  File "/Users/karlkim/chia-network/chia-blockchain/chia/cmds/dao.py", line 8, in <module>
    from chia.cmds.cmds_util import execute_with_wallet
ImportError: cannot import name 'execute_with_wallet' from 'chia.cmds.cmds_util' (/Users/karlkim/chia-network/chia-blockchain/chia/cmds/cmds_util.py)
```


### New Behavior:
DAO cmds works.
```
❯ chia dev sim --root-path $CHIA_SIM_ROOT start -wr
Daemon not started yet
Starting daemon
chia_full_node_simulator: started
chia_wallet: started

❯ chia dao list-proposals -f $voter2_fp -i 2
############################
Proposal ID: 0xb346d8a53b9332f996cbf56608dc05dc534e55e2887983976350a32efda506a0
Status: OPEN
Votes for: 5000
Votes against: 0
------------------------
Proposals have 4 blocks of soft close time.
############################
```

geoffwalmsley committed:
pre-commit fixes for dao_funcs

geoffwalmsley committed:
add reorg tests

AmineKhaldi committed:
Merge main into dao-wallet

geoffwalmsley committed:
revert cat wallet to use generate_signed_transaction

geoffwalmsley committed:
revert get_most_recent_singleton_coin_from_coin_spend in pool_puzzles

geoffwalmsley committed:
move genesis_by_id_or_singleton puzzle into cat_wallet/puzzles

geoffwalmsley committed:
forgot to update clvm tests

geoffwalmsley committed:
lint

geoffwalmsley committed:
fixes for failing CI tests

geoffwalmsley committed:
mypy for test_dao_wallets

geoffwalmsley committed:
add missing proposal_min_amt into update proposal clsp

geoffwalmsley committed:
catch funding coins as change from proposals

geoffwalmsley committed:
fixes to chialisp for update and minting proposals

geoffwalmsley committed:
rewrite proposal tests

geoffwalmsley committed:
pre-commit fixes

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
fix WSM imports after merging main

altendky committed:
Create config.py

geoffwalmsley committed:
p2_singleton_aggregator fixes and tests

geoffwalmsley committed:
mypy for clvm tests

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
fix up merge conflicts

geoffwalmsley committed:
add match_hinted_coin method now required by wallets

geoffwalmsley committed:
add override_memos to GSTOptionalArgs

geoffwalmsley committed:
Merge branch 'dao-wallet' into dao-chialisp-fixes

geoffwalmsley committed:
clvm tests for updated p2_singleton_aggregator puz

geoffwalmsley committed:
add check for distinct voting coin ids to proposal

geoffwalmsley committed:
calculate coin ids and add todo about using ASSERT_EPHEMERAL in future

geoffwalmsley committed:
improve test consistency

geoffwalmsley committed:
more test improvements

geoffwalmsley committed:
Merge branch 'dao-wallet' into dao-chialisp-fixes

geoffwalmsley committed:
attempting to fix hanging untrusted funding test

geoffwalmsley committed:
wait for wallet sync in funding test

geoffwalmsley committed:
farm extra block and increase timeout in funding test

geoffwalmsley committed:
typo

geoffwalmsley committed:
new attempt at funding test

geoffwalmsley committed:
if untrusted funding test get the actual coin record from the peer

geoffwalmsley committed:
try creating the second wallet earlier

geoffwalmsley committed:
Merge branch 'dao-wallet' into dao-chialisp-fixes

geoffwalmsley committed:
comments and remove unused defconst

geoffwalmsley committed:
check for duplicate coin_ids in loop_over_vote_coins

geoffwalmsley committed:
remove unused singleton_inner_puzhash from dao cat tail

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
Merge branch 'dao-wallet' into dao-chialisp-fixes

geoffwalmsley committed:
add sleep to block farming for flaky windows tests

geoffwalmsley committed:
coverage

geoffwalmsley committed:
minor fixes for spend cat proposals

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
Merge branch 'dao-wallet' into dao-chialisp-fixes

geoffwalmsley committed:
remove debug statement

geoffwalmsley committed:
remove unused debug code from wallet

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
add cli tests

geoffwalmsley committed:
pre-commit fix

geoffwalmsley committed:
get config from get_wallet_client in dao_funcs

geoffwalmsley committed:
no asset_ids in dict keys for rpc resp. plus coverage

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
add rpc request type hints

geoffwalmsley committed:
coverage for dao_cat_wallet

geoffwalmsley committed:
coverage for dao_utils

geoffwalmsley committed:
coverage for dao_wallet and misc others

geoffwalmsley committed:
more dao_wallet coverage

geoffwalmsley committed:
last bit of coverage

geoffwalmsley commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2240 @@
+                    new_asset_list = self.dao_info.assets.copy()
+                    new_asset_list.append(asset_id)
+                    dao_info = dataclasses.replace(self.dao_info, assets=new_asset_list)
+                    await self.save_info(dao_info)
```
agreed that this could be slow - we will look to optimise the resync during the next iteration

geoffwalmsley committed:
Merge branch 'dao-wallet' into dao-chialisp-fixes

geoffwalmsley committed:
Dao chialisp fixes (#15997)

Fixes chialisp bugs for DAOs.
- assert my_amount in singleton_aggregator
- ensure that singleton_aggregator has a valid output coin that absorbs
value from merge coins
- check for duplicates in coin_id_list in dao proposal. Raise if present
- add a note that treasury puzzle should use ASSERT_EPHEMERAL when
available to prevent spamming of treasury oracle
- switch to using calculate_coin_id instead of sha256'ing coin info
- added spend_sim tests

geoffwalmsley committed:
updates to singleton store and tests

geoffwalmsley committed:
remove creation of 0 value change coins and fix cat tailhash loop

geoffwalmsley committed:
update wallet for clsp changes, add a test for complex spend proposals

geoffwalmsley committed:
minor fix for cli

geoffwalmsley committed:
update coverage exclusions for new tests

geoffwalmsley committed:
clvm test mistake

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
use TXConfig everywhere

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
add extra_conditions to wallet and rpc

geoffwalmsley committed:
coverage

geoffwalmsley committed:
cover all proposal types in rpc_api test

matt-o-how commented on chia/wallet/puzzles/dao_proposal_validator.clsp:
```diff
@@ -74,11 +74,13 @@
+            (list CREATE_COIN PROPOSAL_EXCESS_PAYOUT_PUZ_HASH (- coin_amount 1) (list (f (r SINGLETON_STRUCT))))
+            conditions
+          )
+      conditions)
```
two thoughts:
- could we have the wallets ignore 0 coins? this would reduce on-chain code. not urgent but just a thought
- please can you fix the indentation here either way?

geoffwalmsley replied:
- the main reasoning for doing it in the chialisp is to prevent proposals creating multiple zero coins which potentially create duplicate condition errors invalidating the whole proposal (particularly for cases where the proposal is spending multiple coins). The other trade off is that while it's more on chain code, it saves the CREATE_COIN cost. I'm not sure which way is more efficient?
- fixed

matt-o-how commented on chia/wallet/puzzles/dao_spend_p2_singleton_v2.clsp:
```diff
@@ -51,9 +51,12 @@
-            (c
-              (list CREATE_COIN P2_SINGLETON_PUZHASH (- total SPEND_AMOUNT) (list (f (r TREASURY_SINGLETON_STRUCT))))
-              output
+            (if (> (- total SPEND_AMOUNT) 0)
```
same comment as before about wallet maybe ignoring this

matt-o-how commented on chia/wallet/puzzles/dao_spend_p2_singleton_v2.clsp:
```diff
@@ -119,9 +122,12 @@
-              (c
-                (list CREATE_COIN P2_SINGLETON_PUZHASH (- (+ total (f (r (f parent_amount_list)))) create_coin_sum) (list (f (r TREASURY_SINGLETON_STRUCT))))
-                CONDITIONS_FOR_THIS_ASSET_TYPE
+              (if (> (- (+ total (f (r (f parent_amount_list)))) create_coin_sum) 0)
```
Same comment as before with regards to 0 value change. Make it the wallet's problem saves on chain space.
But if you are going to keep this here then consider making `(- (+ total (f (r (f parent_amount_list)))) create_coin_sum)` a stored value instead of calculating it twice. Both for readability and for optimisation.

geoffwalmsley replied:
fixed so it's only calculating once

matt-o-how commented on chia/wallet/puzzles/dao_spend_p2_singleton_v2.clsp:
```diff
@@ -186,21 +192,31 @@
+              CONDITIONS
+            )
+          )
+          (for_each_asset_type
```
should we disallow spending coins if there are no `CONDITIONS`?

geoffwalmsley replied:
We need to capture the case where there are TAILHASH_CONDITIONS but not [xch] CONDITIONS. If neither are present then it will just return nil

geoffwalmsley committed:
chialisp edits from matts comments

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+                parent_info = ccparent
+        return parent_info
+
+    async def get_frozen_amount(self) -> uint64:
+        return uint64(0)  # pragma: no cover
```
Does this function need to exist?  It's not part of the wallet protocol and seems like it's unused/pointless

geoffwalmsley replied:
removed

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+            raise ValueError("Could not find any peers to request puzzle and solution from")
+        try:
+            # Get the parent coin spend
+            cs = (await wallet_node.get_coin_state([coin.parent_coin_info], peer, height))[0]
+            parent_spend = await fetch_coin_spend(cs.spent_height, cs.coin, peer)
```
We have a utility function for this probably better to use it: https://github.com/Chia-Network/chia-blockchain/blob/d966f3f9e63aed52dbd73544164202a9f11ce3d2/chia/wallet/util/wallet_sync_utils.py#L428

geoffwalmsley replied:
same as previous comment - it looks to me like we're already using the `fetch_coin_spend` utility?

Quexington replied:
I don't know what you did but this is NOT how it was >:(

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+
+        return
+
+    async def create_tandem_xch_tx(
```
there's a `Wallet.create_tandem_xch_tx` function now that you could probably use.

geoffwalmsley replied:
replaced the old tandem_tx with this one

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+        await self.wallet_state_manager.singleton_store.add_spend(self.id(), eve_coin_spend)
+        return eve_spend_bundle
+
+    def generate_simple_proposal_innerpuz(
```
These `generate_` functions seem like maybe they belong in the dao_utils file

geoffwalmsley replied:
moved to utils

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+        fee: uint64 = uint64(0),
+        push: bool = True,
+        extra_conditions: Tuple[Condition, ...] = tuple(),
+    ) -> Union[SpendBundle, TransactionRecord]:
```
Same comments as in the dao_cat_wallet file about transaction methods only returning `List[TransactionRecord]` and not pushing them automatically.

geoffwalmsley replied:
done

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+        proposed_puzzle_reveal: Program,
+        launcher_coin: Coin,
+        vote_amount: uint64,
+    ) -> SpendBundle:
```
Same comment about transaction methods as above.

geoffwalmsley replied:
done

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+        fee: uint64 = uint64(0),
+        push: bool = True,
+        extra_conditions: Tuple[Condition, ...] = tuple(),
+    ) -> SpendBundle:
```
Same comment about transaction methods as above

geoffwalmsley replied:
done

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+        push: bool = True,
+        self_destruct: bool = False,
+        extra_conditions: Tuple[Condition, ...] = tuple(),
+    ) -> SpendBundle:
```
transaction method comment again

geoffwalmsley replied:
done

Quexington commented on chia/wallet/dao_wallet/dao_wallet.py:
```diff
@@ -0,0 +1,2268 @@
+
+            elif proposal_type == ProposalType.UPDATE:
+                (
+                    _,
```
These underscores come out of nowhere and I can't find where they're defined.  It'd be good to give this constant value a name.

geoffwalmsley replied:
fixed

Quexington commented on chia/wallet/puzzles/p2_conditions_curryable.clsp:
```diff
@@ -0,0 +1,3 @@
+(mod (CONDITIONS_LIST)
```
I think it's slightly dump optimization to just make this `(mod CONDITIONS_LIST` so the compiled output is just `01` but obviously that's a very standard puzzle so this file is certainly not waranted.  I'm not sure why this is necessary, I think the currying is just overhead here instead of just `(q . conditions)`

geoffwalmsley replied:
This file wasn't even being used for anything, so have deleted it

Quexington commented on chia/wallet/singleton.py:
```diff
@@ -14,7 +17,7 @@
 
 
-def get_inner_puzzle_from_singleton(puzzle: Program) -> Optional[Program]:
+def get_inner_puzzle_from_singleton(puzzle: Union[Program, SerializedProgram]) -> Optional[Program]:
```
I think I would prefer `Program.to`ing the call sites instead but idc that much

geoffwalmsley replied:
left as-is for now

Quexington commented on pytest.ini:
```diff
@@ -2,7 +2,7 @@
 log_cli = False
 addopts = --verbose --tb=short -n auto -p no:monitor
-log_level = WARNING
+log_level = INFO
```
Undo this probably

geoffwalmsley replied:
Done

Quexington commented on tools/manage_clvm.py:
```diff
@@ -155,6 +155,9 @@ def from_hex_bytes(cls, hex_bytes: bytes) -> ClvmBytes:
             hash=generate_hash_bytes(hex_bytes=hex_bytes),
         )
 
+    def strip(self) -> ClvmBytes:
+        return ClvmBytes(hex=self.hex.strip(b"\n"), hash=self.hash.strip(b"\n"))
+
```
What's the purpose of this change?  The neline is something we're trying very specifically to commit to so this change is "dangerous" to that idea I think.

geoffwalmsley replied:
I don't know where this came from - have reverted the file to match main

geoffwalmsley committed:
first set of fixes from quex's review

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
more quex fixes

geoffwalmsley committed:
remove push option from wallets

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
move proposal generate funcs to utils, remove tandem tx

geoffwalmsley commented on chia/cmds/dao.py:
```diff
@@ -0,0 +1,958 @@
+    default=None,
+)
+@click.option("-f", "--fingerprint", help="Set the fingerprint to specify which key to use", type=int)
+@click.option("-i", "--wallet-id", help="ID of the DAO wallet which will receive the funds", type=int, required=True)
```
this was a copy/paste error. Fixed

geoffwalmsley commented on chia/cmds/dao.py:
```diff
@@ -0,0 +1,958 @@
+    show_default=True,
+    callback=validate_fee,
+)
+@click.option(
```
added

geoffwalmsley commented on chia/cmds/dao.py:
```diff
@@ -0,0 +1,958 @@
+dao_cmd.add_command(dao_proposal)
+
+
+# TODO: status: how many of your voting coins are locked away vs. spendable, etc.
```
Correct - we don't need to do this, removed the comment

geoffwalmsley commented on chia/cmds/dao_funcs.py:
```diff
@@ -0,0 +1,466 @@
+            mode="existing",
+            tx_config=CMDTXConfigLoader(
+                reuse_puzhash=True,
+            ).to_tx_config(units["chia"], config, fingerprint),
```
fixed

geoffwalmsley commented on chia/wallet/cat_wallet/cat_utils.py:
```diff
@@ -74,6 +74,12 @@ def construct_cat_puzzle(
 
+def construct_cat_puzzlehash_for_inner_puzzlehash(
+    tail_hash: bytes32, inner_puzzle_hash: bytes32
+) -> bytes32:  # pragma: no cover
```
removed

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+            # self.lineage_store = await CATLineageStore.create(self.wallet_state_manager.db_wrapper, self.get_asset_id())
+            # for coin_id, lineage in cat_info.lineage_proofs:
+            #     await self.add_lineage(coin_id, lineage)
+            # await self.save_info(self.cat_info)
```
removed

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+QUOTED_MOD_HASH = calculate_hash_of_quoted_mod_hash(CAT_MOD_HASH)
+
+
+class DAOCATWallet:
```
This was something we considered early on, but I think we had to override a lot of the methods anyway and it became more complex using a subclass than not. @matt-o-how could you fill in the gaps on why we elected not to take the subclass route?

Quexington replied:
I'll resolve this for now, it's something we could look into as an upgrade in the future for better type safety.

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+        # So look for it's parent coin, and get the inner puzzle for it, which should be the same as
+        # the one contained in the lockup.
+        wallet_node: Any = self.wallet_state_manager.wallet_node
+        parent_coin = (await wallet_node.get_coin_state([coin.parent_coin_info], peer, height))[0]
+        parent_spend = await fetch_coin_spend(height, parent_coin.coin, peer)
```
not sure what you meant here - are we not using that utility function already?

Quexington replied:
Hmmmmmmmmmmmmmmmmmmmmm I remember otherwise but github would seem to be gaslighting me into believing this is how it always was and I guess I'm going to let.

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+            # shortcut, works for change
+            lockup_puz = cat_inner
+        else:
+            # TODO: Move this section to dao_utils once we've got the close spend sorted
```
it is sorted. Comment removed

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+        if cat_balance < amount:  # pragma: no cover
+            raise ValueError(f"Insufficient CAT balance. Requested: {amount} Available: {cat_balance}")
+        # get the lockup puzzle hash
+        lockup_puzzle = await self.get_new_puzzle()
```
fixed

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+                        new_cats.append(coin)
+        await self.wallet_state_manager.add_interested_puzzle_hashes([cat_puzzle_hash], [self.id()])
+
+        return txs, new_cats
```
yeah, I agree that the mix of spend bundles and tx records is not ideal. Have made all the wallet functions return a tx record, and moved add_pending_tx to the rpc (except for dao creation which needs to push the tx itself)

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+                    0,
+                    0,
+                    0,
+                    0,
```
done

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+        else:
+            full_spend = spend_bundle
+
+        if push:
```
done (see previous comment)

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+        fee: uint64 = uint64(0),
+        push: bool = True,
+        extra_conditions: Tuple[Condition, ...] = tuple(),
+    ) -> SpendBundle:
```
done

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+
+    async def remove_active_proposal(
+        self, proposal_id_list: List[bytes32], fee: uint64 = uint64(0), push: bool = True
+    ) -> SpendBundle:
```
done

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+            balance += cat_balance
+        return uint64(balance)
+
+    async def sign(self, spend_bundle: SpendBundle) -> SpendBundle:
```
done

geoffwalmsley commented on chia/wallet/cat_wallet/dao_cat_wallet.py:
```diff
@@ -0,0 +1,753 @@
+        """Notification from wallet state manager that wallet has been received."""
+        self.log.info(f"DAO CAT wallet has been notified that {coin} was added")
+        # We can't get the inner puzzle for this coin's puzhash because it has the lockup layer.
+        # So look for it's parent coin, and get the inner puzzle for it, which should be the same as
+        # the one contained in the lockup.
```
The approach here is to reconstruct the inner puzzle by taking the list of active votes in the parent's puzzle, and adding the votes found in the solution. We then check that inner puzzle's hash against the incoming coin's puzhash to make sure we've created the puzzle correctly.

Quexington replied:
Okay that makes sense to me, but this comment seems to tell me you're just assuming the parent's inner puzzle is the child's inner puzzle as well which is what I was nervous about so maybe edit/delete this comment.

geoffwalmsley replied:
comment deleted

geoffwalmsley committed:
use consensus_mode=PLAIN for dao tests

Quexington commented on chia/wallet/cat_wallet/cat_wallet.py:
```diff
@@ -783,6 +783,7 @@ async def generate_signed_transaction(
     ) -> List[TransactionRecord]:
         # (extra_delta, tail_reveal, tail_solution)
         cat_discrepancy: Optional[Tuple[int, Program, Program]] = kwargs.get("cat_discrepancy", None)
+        override_memos: Optional[bool] = kwargs.get("override_memos", None)
```
I just want to make sure I have this straight in my head:
* You use this override_memos argument when you are sending a CAT to a DAO
* The format of this CAT is CAT(inner_puzzle)
* The DAO is recieving this CAT at CAT(p2_singleton_of_dao)

Is that right?  Because it seems to me that the DAO wallet could subscribe to its p2_singleton address and then that would be what the CAT normally hints to as well right?

geoffwalmsley replied:
Yes that's right...the trouble is that p2_singleton_of_dao is unique for each CAT's asset id so everyone's DAO wallet would have to subscribe to all the possible p2_singleton addresses for every CAT in existence.

If you're the one sending the funds, you can subscribe to the p2_singleton puzhash, but then nobody else's DAO wallet would know to do that. The only way we have (currently) to make sure everyone can see incoming funds is to use the dao ID as the hint. The only alternative I can see would be for people to manually create a list of CAT's and have their wallet subscribe to just those assets.

Quexington replied:
I'm not quite sure I'm following.  p2_singleton is parameterized by only the DAO's singleton ID.  Sure each cat will have a different FULL puzzle but the inner puzzle will be the same across all.  It's actually less permutations that our existing system because the standard puzzle can have derivations but p2_singleton is always p2_singleton.

geoffwalmsley replied:
yeah this makes sense. I think in an earlier the version the asset ID was being used somewhere that made this not possible, but you're absolutely right and your suggested way works better. Have removed the cat wallet override, and adapted the dao chialisp, drivers and wallet to hint on the p2_singleton. Thanks for your persistence :)

geoffwalmsley committed:
rework funding puzzles to not use treasury id in hint

geoffwalmsley committed:
remove raise from uncurry_lockup to prevent error or gui start

geoffwalmsley committed:
flaky reorg test

geoffwalmsley committed:
flaky test fixes

geoffwalmsley committed:
more test fixes and coverage

geoffwalmsley committed:
cli fixes

geoffwalmsley committed:
dao balances always 1, and fix tx_ids

geoffwalmsley commented on chia/cmds/dao.py:
```diff
@@ -0,0 +1,958 @@
+    required=True,
+)
+@click.option(
+    "-fa",
```
I've removed the multi-letter options. For DAO creation and rule updates I've left only the long-form options available since having a bunch of random letters is more confusing. And since these commands aren't going to be used terribly frequently I feel like this isn't a bad trade off

geoffwalmsley committed:
revert change to decimal display of mojo-only wallets

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
add fee to tail and fix coin release

geoffwalmsley committed:
remove an unused push in dao cat wallet

geoffwalmsley committed:
Merge branch 'main' into dao-wallet and add valid_times to wallets

geoffwalmsley committed:
add fee_for_cat

geoffwalmsley committed:
fix partial vote amounts in proposal creation

geoffwalmsley committed:
convert spend_bundle to tx_id in CLI

geoffwalmsley committed:
vote amount defaults for other cli funcs

geoffwalmsley committed:
add endpoint for dao rules

geoffwalmsley committed:
Merge branch 'main' into dao-wallet

geoffwalmsley committed:
option to include closed props in list_proposals

geoffwalmsley committed:
CLI tweaks

geoffwalmsley committed:
coverage

wjblanke reviewed (approved):
aok
